src/main/java/htsjdk/samtools/CRAMFileReader.java

```diff
@@ -387,20 +388,85 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
-            seekableStream = (SeekableStream) inputStream;
+        } else if (inputStream != null && inputStream instanceof SeekableStream) {
+            // For SeekableStreams that were provided to the reader constructor instead of a File, we
+            // need to prevent CloseableIterators from closing the underlying stream since we can't
+            // reconstitute a SeekableStream from an InputStream. So wrap the underlying SeekableStream
+            // in a DeferredCloseSeekableStream with a no-op close implementation, and defer closing it
+            // until enclosing reader is closed.
+            if (deferredCloseSeekableStream == null) {
+                deferredCloseSeekableStream = new DeferredCloseSeekableStream((SeekableStream) inputStream);
+            }
+            seekableStream = deferredCloseSeekableStream;
         }
         return seekableStream;
     }
 
+    // In order to reuse a SeekableStream multiple times with CloseableIterators (which close the
+    // underlying stream when they're done), we need to wrap the SeekableStream in an object that
+    // has a no-op close implementation so the actual close can be deferred  until the enclosing
+    // reader is closed.
+    private static class DeferredCloseSeekableStream extends SeekableStream {
+        private final SeekableStream delegateStream;
+
+        public DeferredCloseSeekableStream(final SeekableStream delegateStream) {
+            this.delegateStream = delegateStream;
+            if (delegateStream instanceof DeferredCloseSeekableStream) {
+                throw new IllegalArgumentException("ReuseableSeekableStream objects cannot be nested");
+            }
+        }
+
+        public SeekableStream getDelegate() { return delegateStream; }
+
+        @Override
+        public long length() { return delegateStream.length(); }
+
+        @Override
+        public long position() throws IOException { return delegateStream.position(); }
+
+        @Override
+        public void seek(long position) throws IOException { delegateStream.seek(position); }
+
+        @Override
+        public int read() throws IOException { return delegateStream.read(); }
+
+        @Override
+        public int read(byte[] buffer, int offset, int length) throws IOException {
+            return delegateStream.read(buffer, offset, length);
+        }
+
+        @Override
+        public void close() throws IOException {
+            // defer close, and let the caller close the delegate when its ready to by calling
+            // getDelegate().close()
+        }
+
+        @Override
+        public boolean eof() throws IOException { return delegateStream.eof(); }
+
+        @Override
+        public String getSource() { return delegateStream.getSource(); }
+    }
+
     @Override
     public void close() {
+        // if at any point we created a deferredCloseSeekableStream, close the underlying delegate now
+        if (deferredCloseSeekableStream != null) {
```

Comment:
Isn't delegate.close() handled by closing the input stream below?  Do we need to handle this specially or can we just let the close down below do this?

Reply:
There are a couple of constructors that take `File` that don't initialize `inputStream`. Its not really clear to whether this is truly necessary or not, but I think it's safest to keep it. Having said that, I think this close should be done through CloserUtil, so I'm going to push a commit with that change. Let me know if you disagree.